File: shared/python/engine/counter.py
"""
Logic Engine - Counter

Counter with increment/decrement/reset triggers and external state.
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class CounterConfig:
    """Counter configuration"""
    initial_value: int = 0
    min_value: int = 0
    max_value: int = 100
    step: int = 1
    wrap: bool = False      # Wrap around at limits vs clamp
    edge_mode: bool = True  # Trigger on edge vs level


@dataclass
class CounterState:
    """Counter state (externally managed)"""
    value: int = 0
    last_inc: int = 0
    last_dec: int = 0
    last_reset: int = 0


def _apply_limits(value: int, config: CounterConfig) -> int:
    """Apply min/max limits with wrap or clamp"""
    if config.wrap:
        range_size = config.max_value - config.min_value + 1
        if range_size <= 0:
            return value

        while value > config.max_value:
            value -= range_size
        while value < config.min_value:
            value += range_size
    else:
        if value < config.min_value:
            value = config.min_value
        if value > config.max_value:
            value = config.max_value

    return value


def _detect_rising_edge(last_state: int, current: int) -> tuple[bool, int]:
    """Detect rising edge. Returns (edge_detected, new_last_state)"""
    curr_high = 1 if current != 0 else 0
    prev_high = 1 if last_state != 0 else 0
    return (curr_high and not prev_high), curr_high


def counter_init(state: Optional[CounterState] = None,
                 config: Optional[CounterConfig] = None) -> CounterState:
    """Initialize counter state"""
    if state is None:
        state = CounterState()

    state.value = config.initial_value if config else 0
    state.last_inc = 0
    state.last_dec = 0
    state.last_reset = 0

    return state


def counter_reset(state: CounterState, config: CounterConfig) -> None:
    """Reset counter to initial value"""
    state.value = config.initial_value


def counter_update(
    state: CounterState,
    config: CounterConfig,
    inc_trigger: int,
    dec_trigger: int,
    reset_trigger: int
) -> int:
    """
    Update counter based on triggers.

    Args:
        state: Counter state (modified)
        config: Counter configuration
        inc_trigger: Increment trigger
        dec_trigger: Decrement trigger
        reset_trigger: Reset trigger

    Returns:
        Current counter value
    """
    # Check reset first
    if config.edge_mode:
        edge, state.last_reset = _detect_rising_edge(state.last_reset, reset_trigger)
        if edge:
            counter_reset(state, config)
            return state.value
    else:
        if reset_trigger != 0:
            counter_reset(state, config)
            return state.value

    # Check increment
    do_increment = False
    if config.edge_mode:
        do_increment, state.last_inc = _detect_rising_edge(state.last_inc, inc_trigger)
    else:
        do_increment = inc_trigger != 0
        state.last_inc = 1 if inc_trigger != 0 else 0

    if do_increment:
        state.value += config.step
        state.value = _apply_limits(state.value, config)

    # Check decrement
    do_decrement = False
    if config.edge_mode:
        do_decrement, state.last_dec = _detect_rising_edge(state.last_dec, dec_trigger)
    else:
        do_decrement = dec_trigger != 0
        state.last_dec = 1 if dec_trigger != 0 else 0

    if do_decrement:
        state.value -= config.step
        state.value = _apply_limits(state.value, config)

    return state.value


def counter_increment(state: CounterState, config: CounterConfig) -> int:
    """Direct increment (bypasses triggers)"""
    state.value += config.step
    state.value = _apply_limits(state.value, config)
    return state.value


def counter_decrement(state: CounterState, config: CounterConfig) -> int:
    """Direct decrement (bypasses triggers)"""
    state.value -= config.step
    state.value = _apply_limits(state.value, config)
    return state.value


def counter_set_value(state: CounterState, config: CounterConfig, value: int) -> None:
    """Set counter value directly"""
    state.value = _apply_limits(value, config)


def counter_is_at_min(state: CounterState, config: CounterConfig) -> bool:
    """Check if counter is at minimum"""
    return state.value <= config.min_value


def counter_is_at_max(state: CounterState, config: CounterConfig) -> bool:
    """Check if counter is at maximum"""
    return state.value >= config.max_value

File: shared/engine/counter.c
/**
 * @file counter.c
 * @brief Logic Engine - Counter Implementation
 */

#include "counter.h"
#include <string.h>

/*============================================================================
 * Helper Functions
 *============================================================================*/

static int32_t apply_limits(int32_t value, const Counter_Config_t* config)
{
    if (config->wrap) {
        /* Wrap around */
        int32_t range = config->max_value - config->min_value + 1;
        if (range <= 0) return value;

        while (value > config->max_value) {
            value -= range;
        }
        while (value < config->min_value) {
            value += range;
        }
    } else {
        /* Clamp */
        if (value < config->min_value) value = config->min_value;
        if (value > config->max_value) value = config->max_value;
    }
    return value;
}

static bool detect_rising_edge(uint8_t* last_state, int32_t current)
{
    bool current_high = (current != 0);
    bool last_high = (*last_state != 0);
    *last_state = current_high ? 1 : 0;
    return current_high && !last_high;
}

/*============================================================================
 * Counter Functions
 *============================================================================*/

void Counter_Init(Counter_State_t* state, const Counter_Config_t* config)
{
    if (!state) return;
    memset(state, 0, sizeof(Counter_State_t));

    if (config) {
        state->value = config->initial_value;
    }
}

void Counter_Reset(Counter_State_t* state, const Counter_Config_t* config)
{
    if (!state) return;

    state->value = config ? config->initial_value : 0;
    /* Don't reset edge detection state */
}

int32_t Counter_Update(Counter_State_t* state, const Counter_Config_t* config,
                       int32_t inc_trigger, int32_t dec_trigger,
                       int32_t reset_trigger)
{
    if (!state || !config) return 0;

    /* Check reset first */
    if (config->edge_mode) {
        if (detect_rising_edge(&state->last_reset, reset_trigger)) {
            Counter_Reset(state, config);
            return state->value;
        }
    } else {
        if (reset_trigger != 0) {
            Counter_Reset(state, config);
            return state->value;
        }
    }

    /* Check increment */
    bool do_increment = false;
    if (config->edge_mode) {
        do_increment = detect_rising_edge(&state->last_inc, inc_trigger);
    } else {
        do_increment = (inc_trigger != 0);
        state->last_inc = (inc_trigger != 0) ? 1 : 0;
    }

    if (do_increment) {
        state->value += config->step;
        state->value = apply_limits(state->value, config);
    }

    /* Check decrement */
    bool do_decrement = false;
    if (config->edge_mode) {
        do_decrement = detect_rising_edge(&state->last_dec, dec_trigger);
    } else {
        do_decrement = (dec_trigger != 0);
        state->last_dec = (dec_trigger != 0) ? 1 : 0;
    }

    if (do_decrement) {
        state->value -= config->step;
        state->value = apply_limits(state->value, config);
    }

    return state->value;
}

int32_t Counter_GetValue(const Counter_State_t* state)
{
    return state ? state->value : 0;
}

void Counter_SetValue(Counter_State_t* state, const Counter_Config_t* config,
                      int32_t value)
{
    if (!state) return;

    state->value = config ? apply_limits(value, config) : value;
}

int32_t Counter_Increment(Counter_State_t* state, const Counter_Config_t* config)
{
    if (!state || !config) return 0;

    state->value += config->step;
    state->value = apply_limits(state->value, config);
    return state->value;
}

int32_t Counter_Decrement(Counter_State_t* state, const Counter_Config_t* config)
{
    if (!state || !config) return 0;

    state->value -= config->step;
    state->value = apply_limits(state->value, config);
    return state->value;
}

bool Counter_IsAtMin(const Counter_State_t* state, const Counter_Config_t* config)
{
    if (!state || !config) return false;
    return state->value <= config->min_value;
}

bool Counter_IsAtMax(const Counter_State_t* state, const Counter_Config_t* config)
{
    if (!state || !config) return false;
    return state->value >= config->max_value;
}
